### DISCUSSION TITLE
能否使用这个库解析scrcpy获取到的原始数据呢？

### DISCUSSION THREAD

**Discussion by nnnpa31** [yapingcat/gomedia#49] [Q&A] [answered]
[scrcpy](https://github.com/Genymobile/scrcpy)：Android投屏工具，通过推送scrcpy-server.jar到手机上并执行，使用ADB的端口转发建立TCP连接，将手机屏幕数据一帧帧传输出来。
手机上的服务端启动TCP连接后，客户端应依次执行以下步骤：

- 使用adb forward将手机上的服务映射到电脑的端口
- 与服务器建立一个TCP连接
- 读取1字节长度，用于验证连接是否建立
- 验证成功后，再建立一个TCP连接，作为操作通道
- 读取64字节长度，以utf-8编码转换为字符串后即为手机设备名称等相关信息
- 读取4字节长度，以小端模式分为两个2字节的数据，即为手机分辨率

随后，开启循环读取，每次读取最大0x10000的长度，读取到的数据包起始是0x000001，根据特征判断是h264的nalu裸流。

每个TCP包都可以确定是按时间顺序的一帧，可以按照接收到数据包的真实时间来标记每帧的时间戳。

如何把接收到的h264裸流包装成http-flv或者rtsp等其他形式呢？我的目的是把它通过websocket传输到网页端页面，使用video标签渲染。

**Comment by yapingcat** [comment, marked as answer]
> 将手机屏幕数据一帧帧传输出来

通过什么协议传输的？

> 使用了ffmpeg作为解码库

这个解码是指 h264解码成yuv 这种吗？如果是这种情况，gomedia无能为力
如果你说的解码是指demux 流媒体协议或者文件容器的，那就是第一个问题了，数据是通过什么协议传输的？

**Reply by nnnpa31** [reply]
> 通过什么协议传输的？

似乎是scrcpy的私有协议。我在[Python版的Client实现](https://github.com/razumeiko/py-android-viewer/blob/83f6170a61b6bbc3f20beb57b95a05c282bad7f1/viewer.py#L133)中参考到如下做法：
- 在socket连接中读取0x10000字节的h264 raw数据：raw = socket.recv(0x10000)
- 创建一个AVCodecContext：codec = av.codec.CodecContext.create('h264', 'r')
- 直接把raw数据喂给AVCodecContext：codec.parse(raw)

随后，他们就获得了AVPacket。
我无法在Go中复现对应的步骤，主要卡在创建AVCodecContext和Parse这步。

另：
> 如果你说的解码是指demux 流媒体协议或者文件容器的

获取到AVPacket这一步，似乎就能与gomedia目前可以做的连接起来了。
我的下一步计划是，将它包装为流媒体发送给基于video标签的前端（这是后话），所以不需要decode。

**Reply by yapingcat** [reply]
> > 通过什么协议传输的？
> 
> 似乎是scrcpy的私有协议。我在[Python版的Client实现](https://github.com/razumeiko/py-android-viewer/blob/83f6170a61b6bbc3f20beb57b95a05c282bad7f1/viewer.py#L133)中参考到如下做法：
> 
> * 在socket连接中读取0x10000字节的h264 raw数据：raw = socket.recv(0x10000)
> * 创建一个AVCodecContext：codec = av.codec.CodecContext.create('h264', 'r')
> * 直接把raw数据喂给AVCodecContext：codec.parse(raw)
> 
> 随后，他们就获得了AVPacket。 我无法在Go中复现对应的步骤，主要卡在创建AVCodecContext和Parse这步。
> 
> 另：
> 
> > 如果你说的解码是指demux 流媒体协议或者文件容器的
> 
> 获取到AVPacket这一步，似乎就能与gomedia目前可以做的连接起来了。 我的下一步计划是，将它包装为流媒体发送给基于video标签的前端（这是后话），所以不需要decode。

从描述来看，scrcpy 并没有真正去解码h264.
那么我的理解就是，当前的问题应该就是如何把scrcpy私有协议转成标准的流媒体协议，以提供给浏览器或者其他播放器播放。
不知道我的理解对不对，你可以补充

如果我的理解没有错误，就需要明确一下几个问题
1. 私有协议如何传输音视频的？
    一帧h264都封装在0x10000大小的数据包中？那么其他编码格式的视频数据呢。
   这种传输方式如何区分是音频数据还是视频数据呢？
   还有编码格式如何协商的？只支持h264？
2. 时间戳如何获取？
   scrcpy 私有协议没有传输时间戳？

**Reply by nnnpa31** [reply]
经过重新阅读和学习Python代码，我能继续提供如下现象：

scrcpy在手机端为服务端，经过adb映射后相当于一个单纯的TCP连接。
在启动服务端时，传入参数“8000000”作为码率，"h264"作为编码格式（其它无关参数已省去）
adb映射后可通过127.0.0.1:8081进行连接。

建立连接后，行为依次如下：
- 读取1字节长度，用于验证连接是否建立
- 读取64字节长度，以utf-8编码转换为字符串后即为手机设备名称等相关信息
- 读取4字节长度，以小端模式分为两个2字节的数据，即为手机分辨率

随后，将连接设置为非阻塞模式，循环获取数据并存入队列（变量名称：raw_h264，我根据这个变量名称推测它就是原始h264流）
因为连接已被设定为非阻塞模式，因此每次获取的数据并不一定就是0x10000的大小，可能更小，一次性最大获取这么多。
将数据依次喂给自定义的AVCodecContext，经过序列化得到AVPacket。
将AVPacket送入解码器，获取一帧画面。
前端显示。

至于我的问题，主要在于，我也无法确定这获取到的所谓raw_h264究竟如何解析、有没有时间戳。

然后整理一下上面的几个问题
> 1.私有协议如何传输音视频的？

只有视频，不支持传输音频。参数不填写的默认情况下是h264，暂不知道是否支持其它编码格式。循环获取的数据包只是一次最大0x10000，不能确定一次能获取完整的一帧。

> 2.时间戳如何获取？

似乎没有。因为是按帧的顺序传出的，实时的视频流。

scrcpy这套系统的主要目的是将手机投屏到电脑上，对延时性能的要求比较高，因此可能舍弃了时间戳，转而直接渲染出完整帧并直接显示。

**Reply by yapingcat** [reply]
> 经过重新阅读和学习Python代码，我能继续提供如下现象：
> 
> scrcpy在手机端为服务端，经过adb映射后相当于一个单纯的TCP连接。 在启动服务端时，传入参数“8000000”作为码率，"h264"作为编码格式（其它无关参数已省去） adb映射后可通过127.0.0.1:8081进行连接。
> 
> 建立连接后，行为依次如下：
> 
> * 读取1字节长度，用于验证连接是否建立
> * 读取64字节长度，以utf-8编码转换为字符串后即为手机设备名称等相关信息
> * 读取4字节长度，以小端模式分为两个2字节的数据，即为手机分辨率
> 
> 随后，将连接设置为非阻塞模式，循环获取数据并存入队列（变量名称：raw_h264，我根据这个变量名称推测它就是原始h264流） 因为连接已被设定为非阻塞模式，因此每次获取的数据并不一定就是0x10000的大小，可能更小，一次性最大获取这么多。 将数据依次喂给自定义的AVCodecContext，经过序列化得到AVPacket。 将AVPacket送入解码器，获取一帧画面。 前端显示。
> 
> 至于我的问题，主要在于，我也无法确定这获取到的所谓raw_h264究竟如何解析、有没有时间戳。
> 
> 然后整理一下上面的几个问题
> 
> > 1.私有协议如何传输音视频的？
> 
> 只有视频，不支持传输音频。参数不填写的默认情况下是h264，暂不知道是否支持其它编码格式。循环获取的数据包只是一次最大0x10000，不能确定一次能获取完整的一帧。
> 
> > 2.时间戳如何获取？
> 
> 似乎没有。因为是按帧的顺序传出的，实时的视频流。
> 
> scrcpy这套系统的主要目的是将手机投屏到电脑上，对延时性能的要求比较高，因此可能舍弃了时间戳，转而直接渲染出完整帧并直接显示。

如果没有时间戳，只能解析h264 sps里面的vui，通过vui 里面的 num_units_in_tick 和time_scale  计算帧率了

**Reply by nnnpa31** [reply]
帧速率是在server端启动时通过命令行参数定义好的，默认值是30。但实际上，通常根据手机性能，这个帧率是会波动的。

> 当前的问题应该就是如何把scrcpy私有协议转成标准的流媒体协议，以提供给浏览器或者其他播放器播放。

您总结的非常准确，谢谢，这就是我想寻求的帮助。

我尝试把TCP获取到的数据流存储并通过HexEditor查看解析 ~发现并无规律可循~，每个TCP数据包的起始字节都是0x00001，符合nalu数据的起始码特征。

查阅其他资料与相似实现，这些数据有可能是”[h264的nalu数据](https://www.jianshu.com/p/06f4b8919991)“。
我猜，或许这些数据是并不需要进行私有协议的相关处理的，因为scrcpy似乎并没有为这些数据做包装。
即便是官方实现，也只是直接把它喂给ffmpeg的avcodec。

如果这些数据属于”h264的nalu“，那么是否可以直接使用gomedia将其包装呢？

**Reply by yapingcat** [reply]
> 帧速率是在server端启动时通过命令行参数定义好的，默认值是30。但实际上，通常根据手机性能，这个帧率是会波动的。
> 
> > 当前的问题应该就是如何把scrcpy私有协议转成标准的流媒体协议，以提供给浏览器或者其他播放器播放。
> 
> 您总结的非常准确，谢谢，这就是我想寻求的帮助。
> 
> 我尝试把TCP获取到的数据流存储并通过HexEditor查看解析 ~发现并无规律可循~，每个TCP数据包的起始字节都是0x00001，符合nalu数据的起始码特征。
> 
> 查阅其他资料与相似实现，这些数据有可能是”[h264的nalu数据](https://www.jianshu.com/p/06f4b8919991)“。 我猜，或许这些数据是并不需要进行私有协议的相关处理的，因为scrcpy似乎并没有为这些数据做包装。 即便是官方实现，也只是直接把它喂给ffmpeg的avcodec。
> 
> 如果这些数据属于”h264的nalu“，那么是否可以直接使用gomedia将其包装呢？

可以，但是前提你需要自己计算出时间戳，如果固定是30fps，时间戳可以从0开始，按照33，33，34这样循环累加时间戳

**Reply by nnnpa31** [reply]
> 可以，但是前提你需要自己计算出时间戳，如果固定是30fps，时间戳可以从0开始，可以按照33，33，34这样累加时间戳

时间戳问题的话，我想它也可以直接按接到数据包的真实时间进行累加，因为它是一帧一帧发送并立刻显示的。scrcpy官方说延迟一般不超过1帧，所以应该没有乱序的包。

我手动添加时间戳信息后，似乎就应该做将它封装的步骤了，但看了example里似乎没有像这样处理裸流的案例，请问您的example中哪个文件能为我实现这个步骤提供参考吗？

**Reply by yapingcat** [reply]
> > 可以，但是前提你需要自己计算出时间戳，如果固定是30fps，时间戳可以从0开始，可以按照33，33，34这样累加时间戳
> 
> 时间戳问题的话，我想它也可以直接按接到数据包的真实时间进行累加，因为它是一帧一帧发送并立刻显示的。scrcpy官方说延迟一般不超过1帧，所以应该没有乱序的包。
> 
> 我手动添加时间戳信息后，似乎就应该做将它封装的步骤了，但看了example里似乎没有像这样处理裸流的案例，请问您的example中哪个文件能为我实现这个步骤提供参考吗？

可以参考 [example_mux_ps_h264.go](https://github.com/yapingcat/gomedia/blob/main/example/example_mux_ps_h264.go) ,这个例子是读取裸h264 生成ps流，如果要生成其他格式 muxer替换成其他就可以了

**Reply by nnnpa31** [reply]
已经成功把之前存储的h264流转换成mp4文件了，下一步我就自行编写http-flv的包装了。

作为一个完全不懂音视频相关知识的，强行来开发相关内容的小白，您在对话中引导我对相关内容进行了解和解析，并且让我学会了很多，解决了我目前的这个需求。

非常感谢您的帮助！
